python/graphstorm/model/node_glem.py

```diff
@@ -0,0 +1,311 @@
+        self.gnn = GSgnnNodeModel(alpha_l2norm)
+        self.training_lm = not em_order_gnn_first
+
+    def init_optimizer(self, lr, sparse_optimizer_lr, weight_decay, lm_lr=None):
+        """Initialize optimzer, which will be stored in self.lm._optimizer, self.gnn._optimizer
+        """
+        sparse_params = self.gnn.get_sparse_params()
+        if len(sparse_params) > 0:
+            emb_optimizer = dgl.distributed.optim.SparseAdam(sparse_params, lr=sparse_optimizer_lr)
+            sparse_opts = [emb_optimizer]
+        else:
+            sparse_opts = []
+
+        dense_params = self.gnn.get_dense_params() + self.lm.get_dense_params()
+        if len(dense_params) > 0:
+            optimizer = th.optim.Adam(dense_params, lr=lr,
+                                      weight_decay=weight_decay)
+            dense_opts = [optimizer]
+        else:
+            dense_opts = []
+
+        lm_params = self.lm.get_lm_params()
+        if len(lm_params) > 0:
+            lm_optimizer = th.optim.Adam(lm_params, \
+                                         lr=lm_lr if lm_lr is not None else lr,
+                                         weight_decay=weight_decay)
+            lm_opts = [lm_optimizer]
+        else:
+            lm_opts = []
+        self._optimizer = GSOptimizer(dense_opts=dense_opts,
+                                      lm_opts=lm_opts,
+                                      sparse_opts=sparse_opts)
+
+    def create_optimizer(self):
+        return self._optimizer
+
+    def save_model(self, model_path):
+        """Save either the LM and GNN models.
+        `training_lm` flag determine which actively training model to save
+        """
+        self.lm.save_model(os.path.join(model_path, 'LM'))
+        self.gnn.save_model(os.path.join(model_path, 'GNN'))
+
+    def restore_model(self, restore_model_path, model_layer_to_load=None):
+        """Restore models from checkpoints."""
+        self.lm.restore_model(os.path.join(restore_model_path, 'LM'), model_layer_to_load)
+        self.gnn.restore_model(os.path.join(restore_model_path, 'GNN'), model_layer_to_load)
+
+    def set_node_input_encoder(self, encoder):
+        """Set the node input LM encoder for lm, shared with gnn. 
+        """
+        self.lm.set_node_input_encoder(encoder)
+
+    def set_gnn_encoder(self, gnn_encoder):
+        """Set gnn encoder only for gnn.
+        """
+        self.gnn.set_gnn_encoder(gnn_encoder)
+
+    @property
+    def gnn_encoder(self):
+        """Alias for accessing the gnn_encoder"""
+        return self.gnn.gnn_encoder
+
+    def set_decoder(self, decoder):
+        """Set the same decoder for both, since lm needs to be able to 
+        predict node labels during training
+        """
+        self.lm.set_decoder(deepcopy(decoder))
+        self.gnn.set_decoder(decoder)
+
+    def set_loss_func(self, loss_fn):
+        """Set the final loss function based on the node task 
+        """
+        self.lm.set_loss_func(loss_fn)
```

Comment:
is it possible that the loss function has parameters?

Reply:
That might be possible. However, the `set_loss_func` calls `GSgnnModle.set_loss_func`, which requires `loss_fn` to be a `nn.Module`. Parameters should be passed when initialize the `loss_fn` Module object.

Reply:
if `loss_fn` contains parameters, do we want the loss functions in LM and GNN shared?

Reply:
Yes. Their `loss_fn` should be identical. This is the supervised loss (cross-entropy) for labeled nodes in GLEM.